### ISSUE TITLE
💅 False positive for `lint/style/useConst`

### ISSUE THREAD

**Issue by Stanzilla** [biomejs/biome#2876]
### Environment information

```bash
CLI:
  Version:                      1.7.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v22.1.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.5.2"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Linter:
  Recommended:                  true
  All:                          false
  Rules:                        a11y/noAutofocus = "off"
                                a11y/noSvgWithoutTitle = "off"
                                a11y/useKeyWithClickEvents = "off"
                                a11y/useValidAnchor = "off"
                                complexity/noBannedTypes = "error"
                                complexity/noUselessThisAlias = "error"
                                complexity/noUselessTypeConstraint = "error"
                                correctness/noPrecisionLoss = "error"
                                correctness/noUnusedVariables = "warn"
                                correctness/useExhaustiveDependencies = "error"
                                correctness/useHookAtTopLevel = "error"
                                style/noParameterAssign = "off"
                                style/useAsConstAssertion = "error"
                                style/useBlockStatements = "off"
                                style/useSelfClosingElements = "off"
                                suspicious/noArrayIndexKey = "off"
                                suspicious/noConsoleLog = "off"
                                suspicious/noExplicitAny = "off"
                                suspicious/noExtraNonNullAssertion = "error"
                                suspicious/noMisleadingInstantiator = "error"
                                suspicious/noUnsafeDeclarationMerging = "error"
                                suspicious/useAwait = "off"

Workspace:
  Open Documents:               0
```

### Rule name

useConst

### Playground link

https://biomejs.dev/playground/?code=YwBvAG4AcwB0ACAAZgBvAHIAbQBhAHQAIAA9ACAAKAB2AGEAcgBpAGEAYgBsAGUATgBhAG0AZQA6ACAAcwB0AHIAaQBuAGcALAAgAHMAbwB1AHIAYwBlADoAIAB1AG4AawBuAG8AdwBuACkAIAA9AD4AIAB7AAoAIAAgAGwAZQB0ACAAZgBvAHIAbQBhAHQATwBiAGoAZQBjAHQAOgAgACgAcwByAGMAOgAgAFIAZQBjAG8AcgBkADwAcwB0AHIAaQBuAGcALAAgAHUAbgBrAG4AbwB3AG4APgApACAAPQA%2BACAAcwB0AHIAaQBuAGcAOwAKAAoAIAAgAGMAbwBuAHMAdAAgAGYAbwByAG0AYQB0AE8AYgBqAGUAYwB0AFAAcgBvAHAAZQByAHQAeQAgAD0AIAAoAHAAcgBvAHAAZQByAHQAeQBOAGEAbQBlADoAIABzAHQAcgBpAG4AZwAsACAAYwBvAG4AdABlAG4AdAA6ACAAdQBuAGsAbgBvAHcAbgApACAAPQA%2BACAAewAKACAAIAAgACAAbABlAHQAIABvAHUAdABwAHUAdAAgAD0AIAAnACcAOwAKAAoAIAAgACAAIABzAHcAaQB0AGMAaAAgACgAdAB5AHAAZQBvAGYAIABjAG8AbgB0AGUAbgB0ACkAIAB7AAoAIAAgACAAIAAgACAAYwBhAHMAZQAgACcAcwB0AHIAaQBuAGcAJwA6AAoAIAAgACAAIAAgACAAIAAgAG8AdQB0AHAAdQB0ACAAPQAgAGAAWwA9AFsAJAB7AGMAbwBuAHQAZQBuAHQAfQBdAD0AXQBgADsACgAgACAAIAAgACAAIAAgACAAYgByAGUAYQBrADsACgAgACAAIAAgACAAIABjAGEAcwBlACAAJwBvAGIAagBlAGMAdAAnADoACgAgACAAIAAgACAAIAAgACAAbwB1AHQAcAB1AHQAIAA9ACAAYAB7ACAAJAB7AGYAbwByAG0AYQB0AE8AYgBqAGUAYwB0ACgAYwBvAG4AdABlAG4AdAAgAGEAcwAgAFIAZQBjAG8AcgBkADwAcwB0AHIAaQBuAGcALAAgAHUAbgBrAG4AbwB3AG4APgApAH0AIAB9AGAAOwAKACAAIAAgACAAIAAgACAAIABiAHIAZQBhAGsAOwAKACAAIAAgACAAIAAgAGMAYQBzAGUAIAAnAHUAbgBkAGUAZgBpAG4AZQBkACcAOgAKACAAIAAgACAAIAAgACAAIABvAHUAdABwAHUAdAAgAD0AIABgACIAIgBgADsACgAgACAAIAAgACAAIAAgACAAYgByAGUAYQBrADsACgAgACAAIAAgACAAIABkAGUAZgBhAHUAbAB0ADoACgAgACAAIAAgACAAIAAgACAAdABoAHIAbwB3ACAAbgBlAHcAIABFAHIAcgBvAHIAKAAKACAAIAAgACAAIAAgACAAIAAgACAAYABVAG4AawBuAG8AdwBuACAAdAB5AHAAZQAgACgAJAB7AHQAeQBwAGUAbwBmACAAYwBvAG4AdABlAG4AdAB9ACkAIABvAGYAIABwAHIAbwBwAGUAcgB0AHkAIAAiACQAewBwAHIAbwBwAGUAcgB0AHkATgBhAG0AZQB9ACIAIQBgACwACgAgACAAIAAgACAAIAAgACAAKQA7AAoAIAAgACAAIAB9AAoACgAgACAAIAAgAHIAZQB0AHUAcgBuACAAYABbACIAJAB7AHAAcgBvAHAAZQByAHQAeQBOAGEAbQBlAC4AcgBlAHAAbABhAGMAZQAoAC8AXABcAC8AZwAsACAAJwAnACkAfQAiAF0AIAA9ACAAJAB7AG8AdQB0AHAAdQB0AH0ALABcAG4AYAA7AAoAIAAgAH0AOwAKAAoAIAAgAGYAbwByAG0AYQB0AE8AYgBqAGUAYwB0ACAAPQAgACgAcwByAGMAOgAgAFIAZQBjAG8AcgBkADwAcwB0AHIAaQBuAGcALAAgAHUAbgBrAG4AbwB3AG4APgApACAAPQA%2BAAoAIAAgACAAIABPAGIAagBlAGMAdAAuAGsAZQB5AHMAKABzAHIAYwApAAoAIAAgACAAIAAgACAALgBtAGEAcAAoACgAcAByAG8AcABlAHIAdAB5ACkAIAA9AD4AIABmAG8AcgBtAGEAdABPAGIAagBlAGMAdABQAHIAbwBwAGUAcgB0AHkAKABwAHIAbwBwAGUAcgB0AHkALAAgAHMAcgBjAFsAcAByAG8AcABlAHIAdAB5AF0AKQApAAoAIAAgACAAIAAgACAALgBqAG8AaQBuACgAJwAnACkAOwAKAAoAIAAgAGMAbwBuAHMAdAAgAG8AdQB0AHAAdQB0ACAAPQAgAGYAbwByAG0AYQB0AE8AYgBqAGUAYwB0ACgAcwBvAHUAcgBjAGUAIABhAHMAIABSAGUAYwBvAHIAZAA8AHMAdAByAGkAbgBnACwAIAB1AG4AawBuAG8AdwBuAD4AKQA7AAoACgAgACAAcgBlAHQAdQByAG4AIABgACQAewB2AGEAcgBpAGEAYgBsAGUATgBhAG0AZQB9ACAAPQAgAHsAXABuACQAewBvAHUAdABwAHUAdAB9AFwAbgB9AGAAOwAKAH0AOwAKAAoAZQB4AHAAbwByAHQAIABkAGUAZgBhAHUAbAB0ACAAZgBvAHIAbQBhAHQAOwAKAA%3D%3D

### Expected result

`let formatObject` is correct here, it can't be const because it gets re-assigned later. 

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by Conaclos** [comment]
It is the intended behavior. Here `formatObject` is not initialized and is assigned once. You can turn the single assignment into an initialization.

**Comment by Stanzilla** [comment]
> It is the intended behavior. Here `formatObject` is not initialized and is assigned once. You can turn the single assignment into an initialization.

How do you mean? I did it like that because of the recursion.

**Comment by Conaclos** [comment]
What preventing you of writing the code like this:

```ts
const format = (variableName: string, source: unknown) => {
  const formatObject = (src: Record<string, unknown>) =>
    Object.keys(src)
      .map((property) => formatObjectProperty(property, src[property]))
      .join('');

  const formatObjectProperty = (propertyName: string, content: unknown) => {
    let output = '';

    switch (typeof content) {
      case 'string':
        output = `[=[${content}]=]`;
        break;
      case 'object':
        output = `{ ${formatObject(content as Record<string, unknown>)} }`;
        break;
      case 'undefined':
        output = `""`;
        break;
      default:
        throw new Error(
          `Unknown type (${typeof content}) of property "${propertyName}"!`,
        );
    }

    return `["${propertyName.replace(/\\/g, '')}"] = ${output},\n`;
  };

  const output = formatObject(source as Record<string, unknown>);

  return `${variableName} = {\n${output}\n}`;
};

export default format;
```

**Comment by Stanzilla** [comment]
> What preventing you of writing the code like this:
> 
> ```ts
> const format = (variableName: string, source: unknown) => {
>   const formatObject = (src: Record<string, unknown>) =>
>     Object.keys(src)
>       .map((property) => formatObjectProperty(property, src[property]))
>       .join('');
> 
>   const formatObjectProperty = (propertyName: string, content: unknown) => {
>     let output = '';
> 
>     switch (typeof content) {
>       case 'string':
>         output = `[=[${content}]=]`;
>         break;
>       case 'object':
>         output = `{ ${formatObject(content as Record<string, unknown>)} }`;
>         break;
>       case 'undefined':
>         output = `""`;
>         break;
>       default:
>         throw new Error(
>           `Unknown type (${typeof content}) of property "${propertyName}"!`,
>         );
>     }
> 
>     return `["${propertyName.replace(/\\/g, '')}"] = ${output},\n`;
>   };
> 
>   const output = formatObject(source as Record<string, unknown>);
> 
>   return `${variableName} = {\n${output}\n}`;
> };
> 
> export default format;
> ```

Ah yes, I just figured my case was worth a report, but you are right, it does make sense!
